### ISSUE TITLE
[Messenger] Remove StackInterface

### ISSUE THREAD

**Issue by brunowowk** [symfony/symfony#31179]
**Description**  
We should drop the Stack concept and reintroduce $next in MiddlewareInterface::handle, like in the good old days:
```
interface MiddlewareInterface
{
    /**
     * @param callable|NextInterface $next
     */
    public function handle(Envelope $envelope, callable $next): Envelope;
}

/**	
 * @internal	
 */	
interface NextInterface	
{	
    public function __invoke(Envelope $envelope): void;	
}
```
**Update:** New proposal on #31185

My case agains the Stack is:

1. **It makes middlewares harder to test**

Instead of calling our Middlewares with a simple assertion function we must create a Stack.
When trying to use a simple Stack spy we have to mock both the next() and handle() methods

With the Stack:
```
$stackSpy = $this->prophesize(StackInterface::Class);
$stackSpy->next()->shouldBeCalledOnce()->willReturn($stackSpy);
$stackSpy->handle($envelope, $stackSpy)->shouldBeCalledOnce()->willReturn($envelope);

$myMiddleware->handle($envelope, $stackSpy->reveal());
```

Without the stack:
```
$nextSpy = $this->prophesize(NextInterface::Class);
$nextSpy->__invoke($envelope)->shouldBeCalledOnce()->willReturn($envelope);

$myMiddleware->handle($nextSpy->reveal());
```

2. **It introduces non-determinism in the bus**

Because of the iterator that lives inside the stack we cannot use middlewares to control the bus flow without getting into weird behaviours.

For instance, if not for the stack, the Retry mechanism (#30557, #27215) could be implemented as a Middleware in a trivial way, with the added benefit of being able to retry from specific positions in the pipeline:
```
    public function handle(Envelope $envelope, callable $next) : Envelope
    {
        while ($this->tries-- > 0) {
            try {
                return $next($envelope);
            } catch (\Throwable $t) {
            }
        }
        
        throw $t;
    }
```
If we try this with the Stack it's internal iterator will have advanced, making the middlewares pipeline non-reproducible.

3. **It makes middlewares more verbose**

```
$stack->next()->handle($envelope, $stack);
```
vs
```
$next($envelope);
```

**Comment by nicolas-grekas** [comment]
See #28943 and #29006
TL;DR, good old $next callbacks force doubling the call stack and hurt e.g. debugging

> It introduces non-determinism in the bus

only if you write buggy code. Here is an example that provides what you describe:
```
$next = $stack->next();
$next(...);
$next(...); //retry
```


> It makes middlewares harder to test
> It makes middlewares more verbose

the forced callback indirection that $next callbacks incur is a bigger issue - at least that's the reasoning that led to #28943 and #29006

**Comment by brunowowk** [comment]
>TL;DR, good old $next callbacks force doubling the call stack and hurt e.g. debugging

That's a valid point, but the stack still adds some noise and hurts debugging while also hurting readability and testability.


>only if you write buggy code. Here is an example that provides what you describe:
>
>```
>$next = $stack->next();
>$next->handle(...);
>$next->handle(...); //retry
>```

That code would not work as expected :/
That's because the next handlers on the stack will call $stack->next(), which advances the stack iterator, so even though we froze our reference to the next on the stack, using it like so would case some middlewares in the stack to be skipped when retrying

BTW this behavior is my main motivation for asking for this change because it's not transparent for Middleware developers that the stack behaves this way. IMO this is a dangerous abstraction leakage that could give birth to hard to find bugs when using Messenger.


> the forced callback indirection that $next callbacks incur is a bigger issue

But the callback indirection is restricted to the moment the middleware pipeline is assembled. The $stack->next()->handle() pattern, on the other hand, will be pervasive throughout the whole ecosystem, incurring additional friction on testing and increasing verbosity of all Middlewares.

**Comment by nicolas-grekas** [comment]
Thanks for the explanation about retries. That's a very legit issue.
I opened #31204 to address this issue (keeping the stack as the combined benefit of lazyness+simpler callstack is desired)
